Build fails in environments without network connection

### What happened?

`make WHAT=cmd/kubelet` fails in build environments without network connection due to missing in-tree vendored dependency `go.uber.org/automaxprocs v1.5.2`.

```
$ make WHAT=cmd/kubelet
go version go1.20.6 linux/amd64
go: downloading go.uber.org/automaxprocs v1.5.2
ncpu/main.go:23:2: unrecognized import path "go.uber.org/automaxprocs": https fetch: Get "https://go.uber.org/automaxprocs?go-get=1": dial tcp: lookup go.uber.org on 127.0.1.1:53: server misbehaving
!!! [0804 15:06:39] Call tree:
!!! [0804 15:06:39]  1: /home/etung/kubernetes/hack/lib/golang.sh:843 kube::golang::setup_env(...)
!!! [0804 15:06:39]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0804 15:06:39] Call tree:
!!! [0804 15:06:39]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:92: all] Error 1

```

### What did you expect to happen?

Build to succeed with `make WHAT=cmd/kubelet` without needing to download any go dependencies.

### How can we reproduce it (as minimally and precisely as possible)?

In a network-less build environment, run `make WHAT=cmd/kubelet`.

### Anything else we need to know?

_No response_

### Kubernetes version

Building off of the v1.28.0-rc.0 tag


### Cloud provider

N/A

### OS version

<details>

```console
$ cat /etc/os-release 
NAME="Fedora Linux"
VERSION="37 (Cloud Edition)"
ID=fedora
VERSION_ID=37
VERSION_CODENAME=""
PLATFORM_ID="platform:f37"
PRETTY_NAME="Fedora Linux 37 (Cloud Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:37"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f37/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=37
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=37
SUPPORT_END=2023-11-14
VARIANT="Cloud Edition"
VARIANT_ID=cloud

```

</details>


### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

N/A

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

--- a/hack/lib/golang.sh
+++ b/hack/lib/golang.sh
@@ -556,20 +556,25 @@ kube::golang::setup_env() {
 
   # This seems to matter to some tools
   export GO15VENDOREXPERIMENT=1
+}
 
+kube::golang::setup_gomaxprocs() {
   # GOMAXPROCS by default does not reflect the number of cpu(s) available
   # when running in a container, please see https://github.com/golang/go/issues/33803
-  if ! command -v ncpu >/dev/null 2>&1; then
-    # shellcheck disable=SC2164
-    pushd "${KUBE_ROOT}/hack/tools" >/dev/null
-    GO111MODULE=on go install ./ncpu
-    # shellcheck disable=SC2164
-    popd >/dev/null
+  if [[ -z "${GOMAXPROCS:-}" ]]; then
+    if ! command -v ncpu >/dev/null 2>&1; then
+      # shellcheck disable=SC2164
+      pushd "${KUBE_ROOT}/hack/tools" >/dev/null
+      GO111MODULE=on go install ./ncpu || echo "Will not automatically set GOMAXPROCS"
+      # shellcheck disable=SC2164
+      popd >/dev/null
+    fi
+    if command -v ncpu >/dev/null 2>&1; then
+      GOMAXPROCS=$(ncpu)
+      export GOMAXPROCS
+      kube::log::status "Set GOMAXPROCS automatically to ${GOMAXPROCS}"
+    fi
   fi
-
-  GOMAXPROCS=${GOMAXPROCS:-$(ncpu)}
-  export GOMAXPROCS
-  kube::log::status "Setting GOMAXPROCS: ${GOMAXPROCS}"
 }
 
 # This will take binaries from $GOPATH/bin and copy them to the appropriate

--- a/hack/make-rules/test-e2e-node.sh
+++ b/hack/make-rules/test-e2e-node.sh
@@ -18,6 +18,7 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/../..
 source "${KUBE_ROOT}/hack/lib/init.sh"
 
 kube::golang::setup_env
+kube::golang::setup_gomaxprocs
 
 # start the cache mutation detector by default so that cache mutators will be found
 KUBE_CACHE_MUTATION_DETECTOR="${KUBE_CACHE_MUTATION_DETECTOR:-true}"

--- a/hack/make-rules/test.sh
+++ b/hack/make-rules/test.sh
@@ -22,6 +22,7 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/../..
 source "${KUBE_ROOT}/hack/lib/init.sh"
 
 kube::golang::setup_env
+kube::golang::setup_gomaxprocs
 
 # start the cache mutation detector by default so that cache mutators will be found
 KUBE_CACHE_MUTATION_DETECTOR="${KUBE_CACHE_MUTATION_DETECTOR:-true}"

